machineconfiguration/v1alpha1/types_machineosbuild.go

```diff
@@ -0,0 +1,237 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=machineosbuilds,scope=Cluster
+// +kubebuilder:subresource:status
+// +openshift:api-approved.openshift.io=https://github.com/openshift/api/pull/1773
+// +openshift:enable:FeatureGate=OnClusterBuild
+// +openshift:file-pattern=0000_80_machineosbuildMARKERS.crd.yaml
+// +kubebuilder:metadata:labels=openshift.io/operator-managed=
+
+// MachineOSBuild describes a build process managed by the MCO
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type MachineOSBuild struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec describes the configuration of the machine os build
+	// +kubebuilder:validation:Required
+	Spec MachineOSBuildSpec `json:"spec"`
+
+	// status describes the lst observed state of this machine os build
+	// +optional
+	Status MachineOSBuildStatus `json:"status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineOSBuildList describes all of the Builds on the system
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type MachineOSBuildList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []MachineOSBuild `json:"items"`
+}
+
+// MachineOSBuildSpec describes user-configurable options as well as information about a build process.
+type MachineOSBuildSpec struct {
+	// machineConfigPool is the pool which the build is for
+	// +kubebuilder:validation:Required
+	MachineConfigPool MachineConfigPoolReference `json:"machineConfigPool"`
```

Comment:
the pool associated with this build object is crucial as it lets the build controller know what pool the image is for. This will then auto-populate the renderedMachineConfig reference and a few other entries in the object so that the build controller has the right inputs and rolls out the image to the right nodes.

Without the MCP ref, the build controller does not know who it is building for (or what it should build)

Reply:
OK good context, just thinking out loud but isn't build and deploy two separate actions? Is there an assumption that as soon as the build is completed it is deployed immediately?

Reply:
Would there be a case and maybe its covered in enhancement where user just builds image with the idea it could be deployed in the future. In that case could the existing mechanism for deploying images be used?

Reply:
the image will be immediately deployed to the nodes. I don't know of a circumstance where it would be advantageous to batch build images. so for now, no separate deploy mechanism exists.

Reply:
Separate actions here make sense to me.  I would expect whatever is telling nodes to "update to this level" to use the MachineOSConfig resource.

In which case, this should be a MachineOSConfig reference.